compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp

```diff
@@ -0,0 +1,529 @@
+  Block *moduleBody = innerModule.getBody();
+  rewriter.setInsertionPointToStart(moduleBody);
+
+  // Create a private function call which is the external function call.
+  MLIRContext *context = rewriter.getContext();
+  FunctionType externalFnCallType = getExternalFunctionCallType(
+      context, loc, inputTypes, resultTypes, otherOperandTypes);
+  func::FuncOp externalFnCall =
+      rewriter.create<func::FuncOp>(loc, externalFnName, externalFnCallType);
+  externalFnCall.setPrivate();
+  externalFnCall->setAttr("llvm.bareptr", rewriter.getBoolArrayAttr(true));
+
+  // Create the executable entry point function.
+  std::string entryPointName = externalFnName.str() + "_entry_point";
+  func::FuncOp entryFn =
+      createEntryPointFn(rewriter, rootOp, entryPointName, externalFnCall,
+                         inputTypes, resultTypes, otherOperandTypes);
+
+  // Create the export operation.
+  rewriter.setInsertionPoint(innerModule);
+  auto exportOp = rewriter.create<IREE::Stream::ExecutableExportOp>(
+      loc, entryPointName, FlatSymbolRefAttr::get(context, entryPointName));
+
+  // Create the body of the export operation.
+  // TODO(MaheshRavishankar): This represents the number of workgroups to use.
+  // Ideally this is somehow exposed to the rewrite mechanism to get the
+  // workload and the number of workgroups.
+  Region &exportOpRegion = exportOp.getRegion();
+  Block *exportOpBody =
+      rewriter.createBlock(&exportOpRegion, exportOpRegion.begin());
+  rewriter.setInsertionPointToStart(exportOpBody);
+  Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+  rewriter.create<IREE::Stream::ReturnOp>(loc, ValueRange{one, one, one});
+  return SymbolRefAttr::get(rewriter.getStringAttr(executableOpName),
+                            SymbolRefAttr::get(entryFn));
+}
+
+// Create the `flow.dispatch` op calling into the executable.
+static IREE::Flow::DispatchOp
+createFlowDispatchOp(PatternRewriter &rewriter, SymbolRefAttr exportOp,
+                     Operation *rootOp, TypeRange resultTypes,
+                     ValueRange resultDynamicDims, ValueRange operands) {
+  Location loc = rootOp->getLoc();
+  SmallVector<Value> operandsVec = llvm::to_vector(operands);
+  SmallVector<Value> operandDynamicDims;
+
+  Type i32Type = rewriter.getI32Type();
+  // Get the dynamic dims for the operands.
+  for (auto operand : operands) {
+    auto tensorType = operand.getType().dyn_cast<RankedTensorType>();
+    if (!tensorType)
+      continue;
+
+    for (auto [index, shape] : llvm::enumerate(tensorType.getShape())) {
+      if (!ShapedType::isDynamic(shape))
+        continue;
+
+      Value dim = rewriter.create<tensor::DimOp>(loc, operand, index);
+      // The value into dispatch is passed in as i32.
+      Value dim_i32 = rewriter.create<arith::IndexCastOp>(loc, i32Type, dim);
+      operandsVec.push_back(dim_i32);
+      operandDynamicDims.push_back(dim);
+    }
+  }
+
+  // Append the result dynamic dims as i32.
```

Comment:
why i32 and not i64? if we have to pick a physical type i64 would be better
index should work just fine, though, and be much more readable through the entire flow (and let the options we have for choosing index -> physical width work properly)
you need to pick a physical width for crossing the C ABI boundary, but not for the stuff in the compiler with flow.dispatch.

(if you hit issues with index/i64 let me know - we should properly handle them)